### ISSUE TITLE
systemd initrd decryption prompt times out after 90 seconds

### ISSUE THREAD

**Issue by lopsided98** [NixOS/nixpkgs#250003]
### Describe the bug

With systemd initrd and an encrypted rootfs, the system enters emergency mode if the decryption password is not entered within 90 seconds. This occurs because systemd device units time out by default after 90 seconds. Additionally, ZFS on LUKS (I haven't tested native ZFS encryption) times out after 60 seconds because of a [hardcoded timeout](https://github.com/NixOS/nixpkgs/blob/2f4e3f6fa553264d3b8bcf4d3561c0134b625afa/nixos/modules/tasks/filesystems/zfs.nix#L144) in `zfs-import-${pool}.service`.

To work around the first issue, I added [`x-systemd.device-timeout=0` to the root filesystem options](https://github.com/lopsided98/nixos-config/blob/83320b0cc963285220684dc786d29b3cff1ced19/machines/p-3400/default.nix#L9). To fix ZFS, I added the decrypted rootfs device unit as a dependency of `zfs-import-root.service`, so the 60 second timeout doesn't start until the device is decrypted. I also set `JobTimeoutSec=infinity` on the device unit (see [here](https://github.com/lopsided98/nixos-config/blob/83320b0cc963285220684dc786d29b3cff1ced19/machines/HP-Z420/default.nix#L122-L133)).

I don't see an obvious way to turn these workarounds into something that can be automatically configured in nixpkgs, but I think we should find some solution, because the current behavior is unexpected and annoying.

### Notify maintainers

@ElvishJerricco

**Comment by surfaceflinger** [comment]
Can confirm that it's broken with zfs native encryption too

**Comment by ElvishJerricco** [comment]
So this is a tricky problem. For non-ZFS, systemd is *supposed* to handle this. I've simulated the `systemd-cryptsetup-generator` and the `systemd-fstab-generator` and look at this:

```
$ cat etc/fstab
/dev/mapper/virt /foo ext4 defaults 0 0

$ cat etc/crypttab
virt LABEL=phys

$ unshare -U -r -m bash -c 'mount --bind /nix/store ./nix/store; mount --bind /proc ./proc; chroot . /run/current-system/systemd/lib/systemd/system-generators/systemd-fstab-generator /run/systemd/generator /run/systemd/generator.early /run/systemd/generator.late'

$ unshare -U -r -m bash -c 'mount --bind /nix/store ./nix/store; mount --bind /proc ./proc; chroot . /run/current-system/systemd/lib/systemd/system-generators/systemd-cryptsetup-generator /run/systemd/generator /run/systemd/generator.early /run/systemd/generator.late'

$ find run/systemd/
run/systemd
run/systemd/generator
run/systemd/generator/cryptsetup.target.requires
run/systemd/generator/cryptsetup.target.requires/systemd-cryptsetup@virt.service
run/systemd/generator/dev-mapper-virt.device.d
run/systemd/generator/dev-mapper-virt.device.d/40-device-timeout.conf
run/systemd/generator/systemd-cryptsetup@virt.service
run/systemd/generator/local-fs.target.requires
run/systemd/generator/local-fs.target.requires/foo.mount
run/systemd/generator/foo.mount
run/systemd/generator/local-fs.target.wants
run/systemd/generator/local-fs.target.wants/systemd-remount-fs.service
run/systemd/generator/dev-mapper-virt.device.requires
run/systemd/generator/dev-mapper-virt.device.requires/systemd-cryptsetup@virt.service
run/systemd/generator.late
run/systemd/generator.early

$ cat run/systemd/generator/dev-mapper-virt.device.d/40-device-timeout.conf
# Automatically generated by systemd-cryptsetup-generator

[Unit]
JobTimeoutSec=infinity
```

So what's happening here is that the file system `foo.mount` won't be started until `dev-mapper-virt.device` appears, but because of `dev-mapper-virt.device.d/40-device-timeout.conf`, that device will never timeout. The device timeout works correctly because `systemd-cryptsetup@virt.service` requires and is ordered after `dev-disk-by\x2dlabel-phys.device`, so the timeout on `dev-disk-by\x2dlabel-phys.device` will cause cascading failures in the event that the physical device fails to show up.

Now, for those having this problem in this use case: **This means you should set your file system device to `/dev/mapper/foo`, not `/dev/disk/by-whatever/whatever`.** This will ensure that it only times out when the actual physical device fails to appear, not when you fail to enter the passphrase in time. By depending on anything other than `/dev/mapper/foo`, you're failing to get the timeout override that makes this all work.

As for ZFS, the problem is analogous. The import service is not depending on the device that it needs to import, so it's starting too early and assuming the device has failed to appear. In an ideal world, ZFS would have udev rules that makes a device indicative of the pool name only appear once the pool's drives are all available, so that we could order the import service after said device. For now, the best alternative is probably to order the import service after `cryptsetup.target`, but this is not without caveats. For instance, what about users who have `crypttab` devices stored on ZFS zvols? The import service actually needs to come *before* `cryptsetup.target` in that case. Not sure what exactly to do here.

**Comment by lopsided98** [comment]
Thank you for looking into this; I can confirm that this fixes the non-ZFS case. Quickly looking at the code, it seems like it would be relatively simple for systemd to apply the drop in to the `by-uuid` path as well, but then I guess you could still break it by using one of the the `by-id` paths or `/dev/dm-0`.

**Comment by SuperSandro2000** [comment]
I've also implemented https://github.com/lopsided98/nixos-config/blob/master/machines/HP-Z420/default.nix#L121-L134 which I think we should be able to auto generate. If it is not generic solvable, then at least behind an option which may be by default on.

Also we should treat this issue with a bit of priority because I ended up several times in emergency mode when the unlocking failed even after already entering the normal system.

**Comment by ElvishJerricco** [comment]
@SuperSandro2000 setting the timeout to infinity is definitely not correct. The right thing to do is to add systemd orderings on either the correct device units or on `cryptsetup.target`.

**Comment by ElvishJerricco** [comment]
@SuperSandro2000 Don't create custom `.device` units like that. Just order the zfs service against the mapper names, as I described above, instead of `/dev/disk/by-uuid` names. The LUKS logic is already taking care of finding disks by UUID if that's what you care about.

There isn't an option that could be turned on by default, because the disks required differ from system to system. The only common thing that *could* be done by default is ordering after `cryptsetup.target`, but as I said before, this breaks other setups that have LUKS devices *on the zpool*.

**Comment by SuperSandro2000** [comment]
You mean like?

```nix
boot.initrd.systemd.services."zfs-import-root" = let
  zfsPools = [
    "dev-mapper-machine\\x2deins.device"
    "dev-mapper-machine\\x2dzwei.device"
  ];
  in {
    wants = zfsPools;
    after = zfsPools;
  };
```

**Comment by ElvishJerricco** [comment]
Yep. That way you don't need the custom timeout in the `units = ...` stuff.

**Comment by preisi** [comment]
> Now, for those having this problem in the non-ZFS case: **This means you should set your file system device to `/dev/mapper/foo`, not `/dev/disk/by-whatever/whatever`.** This will ensure that it only times out when the actual physical device fails to appear, not when you fail to enter the passphrase in time. By depending on anything other than `/dev/mapper/foo`, you're failing to get the timeout override that makes this all work.

Thank you very much for looking into this issue. However, even with your recommended fix the password prompt times out after 90 seconds (the result being the emergency mode). It seems that mounting the `/boot` partition is the problem as it is the only remaining file system device still being supplied as `/dev/disk/by-uuid/...`. (Also, it fails to mount when NixOS tries to keep on booting after decryption after the timeout).
Then again, `/boot` is not encrypted in my specific case, my setup is rather simple with an unencrypted boot-partition (vfat) + LVM-over-Luks for the remainder (which are supplied via `/dev/mapper/...`).

Do you have any pointers/ideas on how to further debug this or even a solution at hand?

**Comment by ElvishJerricco** [comment]
@preisi Sorry for taking a while to get back to you.

> It seems that mounting the /boot partition is the problem

The `/boot` partition is not something stage 1 handles. If that's causing delays, it's during stage 2, and it's a separate issue.

**Comment by devurandom** [comment]
Will this make it into [24.11](https://github.com/NixOS/nixpkgs/issues/344920)?

**Comment by ElvishJerricco** [comment]
@devurandom What do you mean by "this"? systemd initrd is already available, it's just not enabled by default. We don't have a solution for this issue yet; though I think what we're likely to do is just disable the timeouts altogether by default.

**Comment by devurandom** [comment]
I had read https://github.com/NixOS/nixpkgs/issues/344920#issuecomment-2380315041 and indeed meant "disable the timeouts altogether by default" when I wrote "this".  Thanks!

**Comment by hustlerone** [comment]
I want to add on this that this is an issue with LVM under LUKS. NixOS has no idea that /dev/LVM_VG/Whatever actually depends on some encrypted drive.

**Comment by BerriJ** [comment]
> I want to add on this that this is an issue with LVM under LUKS. NixOS has no idea that /dev/LVM_VG/Whatever actually depends on some encrypted drive.

+1 for this comment.

I'm using LVM under LUKS on my server and this timeout thing is really bothering me. This renders remote unlocking almost useless as I cannot gurantee to SSH into this machine in 90 seconds after booting. 
I already tried using /dev/mapper/* in the filesystem options but the machine still times out after 90 seconds. If someone can give guidance on how to edit this timeout, I'd really appreciate that.

**Comment by ElvishJerricco** [comment]
A workaround would be to add the same infinity timeout that systemd adds to the LUKS dm-crypt mapper.

```
boot.initrd.systemd.units."dev-vg-lv.device" = {
  overrideStrategy = "asDropin";
  text = ''
    [Unit]
    JobTimeoutSec=infinity
  '';
};
```

Or just set the default device timeout to infinity

```
boot.initrd.systemd.extraConfig = ''
  DefaultDeviceTimeoutSec=infinity
'';
```

**Comment by hustlerone** [comment]
And what about cryptsetup failing to start after two retries?

**Comment by ElvishJerricco** [comment]
@hustlerone I have no idea what you're talking about. Is that a different issue?

**Comment by hustlerone** [comment]
Yes

**Comment by hustlerone** [comment]
Fixed it on my own config, putting out a fix shortly.

**Comment by jbiel** [comment]
Thanks for this thread. I ran into this problem (running mirrored ZFS root on 2 LUKS volumes w/ systemd initrd) and so far none of the workarounds mentioned get past the ~90s "emergency mode" problem. I will keep poking and update if I'm able to find a workable solution.

**Comment by hustlerone** [comment]
> [@hustlerone](https://github.com/hustlerone) is your fix still valid?
> 
> I just stumbled across this issue again, as my fix:
> 
>   boot = {
>     initrd.systemd = {
>       enable = true;
>       settings.Manager = ''
>         DefaultDeviceTimeoutSec=infinity
>       '';
>     };
>   };
> 
> does not work anymore due to [#426692](https://github.com/NixOS/nixpkgs/pull/426692)
> 
> Well, I guess I just have to adjust the config accordingly but fixing this properly would be the better way.
> 
> Edit: It has to be an attractionset now:
> 
>   boot = {
>     initrd.systemd = {
>       enable = true;
>       settings.Manager = {
>         DefaultDeviceTimeoutSec = "infinity";
>       };
>     };
>   };

My fix is still valid probably let me check it out

**Comment by hustlerone** [comment]
oh you fixed it nice

**Comment by BerriJ** [comment]
> oh you fixed it nice

Who do you refer to? Is it fixed already?

**Comment by ElvishJerricco** [comment]
@hustlerone I don't think this should have been closed (yet). Although there are various workarounds for the timeout, they aren't documented, and the determination was not made here that those workarounds constitute a sufficient solution.

Offline, I've concluded with other maintainers that these workarounds will *have* to be sufficient, because there is no general solution. But that was not made clear here, and more importantly it still needs to be put into documentation, which I am working on as a part of a systemd stage 1 migration guide.

**Comment by hustlerone** [comment]
Sounds good to me

**Comment by jbiel** [comment]
I just want to say that I appreciate the continued energy that's being put into this! My current tilt at this has been unsuccessful and I'm going to try the latest solution that @hustlerone posted today. Thank you.

**Comment by hustlerone** [comment]
I apologize for forgetting about this issue for 3 months straight (i fixed it on my config and forgot about it)

**Comment by ElvishJerricco** [comment]
@jbiel I think the simplest solution is to just set e.g. `fileSystems."/".options = [ "x-systemd.device-timeout=infinity" ];`

**Comment by jbiel** [comment]
> `fileSystems."/".options = [ "x-systemd.device-timeout=infinity"];`

Unfortunately this does not work for me on NixOS 25.05. I still get a timeout and panic (`You are in emergency mode ... Press enter to continue`) if the LUKS passphrase for the root is not unlocked within ~90 seconds. My rootfs is RAID 1 via ZFS on LUKS volumes, and I'm using `systemd-boot` for the ability to unlock LUKS via SSH.

I'm trying [the solution outlined in this comment](https://github.com/NixOS/nixpkgs/issues/250003#issuecomment-3179786204), but hitting a wall pretty early on due to the option name not being valid; I don't see an equivalent or something close [in that namespace](https://search.nixos.org/options?channel=25.05&from=0&size=50&sort=relevance&type=packages&query=boot.initrd.systemd).

```
       error: The option `boot.initrd.systemd.settings' does not exist. Definition values:
       - In `/etc/nixos/configuration.nix':
           {
             Manager = {
               DefaultDeviceTimeoutSec = "infinity";
             };
           }
```

Some of my config:

```
  fileSystems."/" =
    { device = "rpool/root";
      fsType = "zfs";
      options = [ "x-systemd.device-timeout=infinity"];
    };

  boot.loader.systemd-boot.enable = true;
  boot.loader.efi.canTouchEfiVariables = true;

  boot.initrd.luks.devices = {
    "luks-rpool-0".device = "/dev/disk/by-uuid/UUIDX";
    "luks-rpool-1".device = "/dev/disk/by-uuid/UUIDY";
  };

  boot.initrd = {
    network = {
      ssh = {
        enable = true;
        port = 2222;  # Use a separate port to prevent conflict in your known hosts
        hostKeys = [ "/path/to/host/key" ];
        authorizedKeys = "authorizedKeys";
      };
    };
  };
  boot.initrd.systemd.enable = true;
  boot.initrd.systemd.users.root.shell = "/bin/systemd-tty-ask-password-agent";
  boot.initrd.systemd.network = config.systemd.network;
```

This all works great other than a rootfs timeout being hit if the LUKS passphrase isn't entered in short order. When I'm changing these options it's not taking long to do `nixos-rebuild switch`. It's definitely not building a new `initrd`, and I wonder if that's correct.

**Comment by ElvishJerricco** [comment]
@jbiel ZFS is different, it has its own timeout for importing the pool. Unfortunately I don't think we have a way to configure that timeout right now. You can fix it by setting:

```
boot.initrd.systemd.services.zfs-import-rpool = {
  after = [ "cryptsetup.target" ];
  wants = [ "cryptsetup.target" ];
};
```

The solution using `boot.initrd.systemd.settings.Manager.DefaultDeviceTimeoutSec` wouldn't help in the ZFS case, because ZFS is not using the systemd device timeout; like I said it has its own separate timeout. But for the sake of completeness, the `settings` option style is new in unstable; on 25.05 it would be:

```
boot.initrd.systemd.extraConfig = ''
  DefaultDeviceTmeoutSec=infinity
'';
```

**Comment by jbiel** [comment]
Thanks @ElvishJerricco , that works for my ZFS root!

```
boot.initrd.systemd.services.zfs-import-rpool = {
  after = [ "cryptsetup.target" ];
  wants = [ "cryptsetup.target" ];
};
```

**Comment by stackcoder** [comment]
Thank you all for providing debugging and help on this!

I've put following together, which seems to work for me:

```nix
boot.initrd.luks.devices."luks-rpool-XXX" = {
  # Default stops after two failed password attempts and drops to emergency shell
  crypttabExtraOpts = [ "tries=0" ];
};

# ZFS pool import runs in parallel and hits the 60s timeout; wait for password entry first
boot.initrd.systemd.services.zfs-import-rpool = {
  after = [ "cryptsetup.target" ];
  wants = [ "cryptsetup.target" ];
};
```

tbo, systemd Stage1 still needs further improvement and documentation for more advanced disk setups.
